elizaOS/eliza
# feat: Add image generation support for tweets

# Relates to:





# Risks

Medium. This change:
- Introduces new dependencies that need to be maintained
- Adds image generation which could affect tweet posting reliability
- Involves file system operations for image handling
- Could impact API rate limits with Twitter when posting media

# Background

## What does this PR do?

This PR adds configurable image generation capabilities to the Twitter client:
- Integrates @ai16z/plugin-image-generation for image creation
- Adds configurable probability for image generation (0-100%)
- Implements image buffer handling and data URL processing
- Adds fallback to text-only tweets if image generation fails
- Improves logging and scheduling visibility for tweets
- Updates dependencies to support new functionality

## What kind of change is this?

Features (non-breaking change which adds functionality):
- New image generation feature
- Enhanced tweet scheduling visibility
- Improved logging and error handling
- New configuration options

# Documentation changes needed?

My changes require a change to the project documentation:
- Document new IMAGE_GEN_CHANCE environment variable
- Add setup instructions for image generation feature
- Document image generation probability configuration
- Update Twitter client usage documentation with new image capabilities



# Testing
- Testing different IMAGE_GEN_CHANCE values (0%, 50%, 100%)
- Verifying image generation and posting with media
- Confirming fallback to text-only tweets when needed
- Checking logging output for scheduling and image generation status

## Where should a reviewer start?
```sh
git checkout twitter/image-gen
pnpm clean
pnpm i
pnpm build
pnpm start
```

### HEURIST Creds
API Key (https://heurist.ai/dev-access)
Heurist Models (https://docs.heurist.ai/dev-guide/supported-models#large-language-models-llms)


## Detailed testing steps
- in .env
    - Enter all Twitter credentials
    - Set `IMAGE_GEN=TRUE`
    - Verify that images are tweeted by setting chance to 100%
    - Set `HEURIST_API_KEY`
    - Set `HEURIST_IMAGE_MODEL`
Otherwise, run `pnpm start` as normal

Right now `IMAGE_GEN=FALSE` is a little quirky 
If `IMAGE_GEN=FALSE` then user should also set `IMAGE_GEN_CHANCE=0`
